ports/servo/glutin_app/window.rs

```diff
@@ -408,56 +408,68 @@ impl Window {
+            ch = (ch as u8 + 96) as char;
+        }
+        let mut event = if let Some(event) = self.last_pressed.replace(None) {
+            event
         } else {
             return;
         };
-
-        self.last_pressed_key.set(None);
-
-        let (key, ch) = if let Some(key) = keyutils::char_to_script_key(ch) {
-            (key, Some(ch))
-        } else {
-            (last_key, None)
-        };
-
-        let modifiers = self.key_modifiers.get();
-        let event = WindowEvent::KeyEvent(ch, key, KeyState::Pressed, modifiers);
-        self.event_queue.borrow_mut().push(event);
+        event.key = Key::Character(ch.to_string());
+        self.event_queue
+            .borrow_mut()
+            .push(WindowEvent::Keyboard(event));
     }
 
     fn toggle_keyboard_modifiers(&self, mods: ModifiersState) {
         self.toggle_modifier(KeyModifiers::CONTROL, mods.ctrl);
         self.toggle_modifier(KeyModifiers::SHIFT, mods.shift);
         self.toggle_modifier(KeyModifiers::ALT, mods.alt);
-        self.toggle_modifier(KeyModifiers::SUPER, mods.logo);
+        self.toggle_modifier(KeyModifiers::META, mods.logo);
     }
 
     fn handle_keyboard_input(
         &self,
+        scancode: u32,
         element_state: ElementState,
-        code: VirtualKeyCode,
+        code: Option<VirtualKeyCode>,
         mods: ModifiersState,
     ) {
+        info!(
+            "winit keyboard input: {}, {:?}, {:?}, {:?}",
+            scancode, element_state, code, mods
+        );
         self.toggle_keyboard_modifiers(mods);
-
-        if let Ok(key) = keyutils::winit_key_to_script_key(code) {
-            let state = match element_state {
-                ElementState::Pressed => KeyState::Pressed,
-                ElementState::Released => KeyState::Released,
-            };
-            if element_state == ElementState::Pressed && keyutils::is_printable(code) {
-                // If pressed and printable, we expect a ReceivedCharacter event.
-                self.last_pressed_key.set(Some(key));
-            } else {
-                self.last_pressed_key.set(None);
-                let modifiers = self.key_modifiers.get();
-                self.event_queue
-                    .borrow_mut()
-                    .push(WindowEvent::KeyEvent(None, key, state, modifiers));
-            }
+        let key = keyutils::get_servo_key_from_winit_key(code);
+        let location = keyutils::get_servo_location_from_winit_key(code);
+        let state = match element_state {
+            ElementState::Pressed => KeyState::Down,
+            ElementState::Released => KeyState::Up,
+        };
+        let event = KeyboardEvent {
+            state,
+            key: key.clone(),
+            code: keyutils::get_servo_code_from_scancode(scancode),
+            location,
+            modifiers: self.key_modifiers.get(),
+            repeat: false,
+            is_composing: false,
+        };
```

Comment:
Maybe introduce `keyutils::keyevent_from_winit() -> KeyboardEvent` instead of having all these get_servo__*_from_winit.